### ISSUE TITLE
linear reorder?

### ISSUE THREAD

**Issue by behdad** [fribidi/fribidi#30]
Should port to using something like https://github.com/fribidi/linear-reorder/blob/master/linear-reorder.c

@khaledhosny did you say you have that?

**Comment by khaledhosny** [comment]
There is #11, but it does not merge since the code it modifies now gone.

**Comment by capr** [comment]
If you can elucidate me please as I can't figure this out: UAX#9 3.4 says that reordering is applied after any word wrapping is applied to the paragraph.  But how can word wrapping be implemented before knowing the order of the text segments are so that their widths can be added and compared to the box width? Looks like a chicken-egg problem to me. Any thoughts?

**Comment by capr** [comment]
...also it looks like the reordering code in fribidi is more complex and also it wants to know the resolved paragraph base dir, while this code doesn't. Can this replace fribidi's reordering code entirely or there's more to it? 

Sorry if these questions sound confused.  The thing is I cannot use fribidi for reordering because fribidi also reorders the contents of the RTL runs which harfbuzz also does, resulting in gibberish, from which I concluded that there's really no way to use fribidi's reordering function with harfbuzz so I have to implement my own. Is this assumption correct?

**Comment by behdad** [comment]
> But how can word wrapping be implemented before knowing the order of the text segments so that their widths can be added and compared to the box width? Looks like a chicken-egg problem to me. Any thoughts?

Wrapping works on the logical order of the text segments, so you don't need to know the reordering results.

**Comment by behdad** [comment]
> ...I concluded that there's really no way to use fribidi's reordering function with harfbuzz so I have to implement my own. Is this assumption correct?

Correct. Try using linear-reorder in your code directly.  We should add a solution to fribidi itself some time.

**Comment by capr** [comment]
@behdad thanks for clearing that up!

**Comment by khaledhosny** [comment]
I opened a new PR for linear reorder, #84, restoring a bit of `fribidi_reorder_runs()` enough to make it compile. Did not test it though.

**Comment by capr** [comment]
After implementing this and seeing the results, it looks like I still need to know the visual order of segments in order to do proper line wrapping. The reason being that I cannot just compute the line width as the sum of the advance widths of the text segments (which is the same before and after reordering). I also need to substract the left side bearing of the first glyph (and the RSB of the last glyph) to get the actual bounding-box width of the line. If I don't take side bearings into account then the text will overflow the box eg. when ending on an "f" on most fonts. Underflowing is not a problem, but overflowing can strike borders etc. It seems that browsers have this problem too, eg. try the code below on chrome, you'll see how the text overflows its box. Any solution to this?

~~~
<style>
span {
	font-size: 100px;
	margin: 0;
	padding: 0;
	border: 1px solid black;
	width: 0px;
}
</style>
<span>
	fff
</span>
~~~

**Comment by khaledhosny** [comment]
I don’t think you should be subtracting side bearing when calculating line width, and what browsers do is what I’d expect and I think is pretty much what any other application is doing.

**Comment by capr** [comment]
Hi Khaled! Thanks for the input. I still don't understand something: when you do line wrapping on a text editor, you need to compute the width of the line which has to exclude the whitespace at the end of the last word (which can be different for each word). But I don't know what the last word is before bidi reordering. It's basically the same problem as before but now with whitespace.

**Comment by capr** [comment]
Here's a concrete example:

The text "لمفاتيح ABC DEF" with RTL base direction renders as "ABC DEF لمفاتيح". When wrapped, DEF goes on the second line (as it should), but the space after "C" should be removed when wrapping.

![1](https://user-images.githubusercontent.com/131660/44721744-8c17ee00-aad3-11e8-8ee5-e4328d1ad148.png)
![2](https://user-images.githubusercontent.com/131660/44721751-90dca200-aad3-11e8-90ba-f688a4774898.png)

I'm just looking for the general pattern/algorithm on which whitespace should be removed when. UAX#9 has the following phrase on section 3.4 but I don't understand it: "In combination with the following rule, this means that trailing whitespace will appear at the visual end of the line (in the paragraph direction)."

**Comment by khaledhosny** [comment]
What I did in the past is:
* apply bidi and reorder the runs, 
* shape,
* save the visual order of each glyph,
* sort glyphs by the logical order,
* break the lines, 
* sort glyphs back in each line by the visual order.

There might be better alternatives, but that was the simplest solution for me.

**Comment by capr** [comment]
The problem in my last message seem to be because the linear reorder algorithm only does L2 from UAX#9 and I also need to do L1 which should move that trailing whitespace away to the end of the line. So sorry for the noise on that. 

On that note, looks like harfbuzz does L4 (mirroring) and it doesn't need L3 (combining-marks-must-come-after-base-char) since the segments are not reversed, so only L1 was missing after all.

**Comment by behdad** [comment]
> Hi Khaled! Thanks for the input. I still don't understand something: when you do line wrapping on a text editor, you need to compute the width of the line which has to exclude the whitespace at the end of the last word (which can be different for each word). But I don't know what the last word is before bidi reordering. It's basically the same problem as before but now with whitespace.

The whitespace to be removed is always the whitespace last in line in the logical direction.  So, again, you don't need to know the order of runs in the line to elide that.

**Comment by mrcasty** [comment]
Hi guys, 

Are there plans to push this change in a near future? I have been waiting for this change like 5 years, and I would merge it locally myself, but the fact that you are also hesitating scares me.

**Comment by behdad** [comment]
Neither of us have much time for fribidi. But if Khaled can use his proposed API in libraqm, I'm happy. That's in https://github.com/fribidi/fribidi/pull/84

**Comment by dov** [comment]
Check commit https://github.com/fribidi/fribidi/commit/9793c192b7086066cf65b509f36871b8e3fe071d .  I believe this issue has been solved. Reopen if not. 

Note that the commit made part of the benchmark test 33x faster.
